### ISSUE TITLE
[BUG] Empty result set for k-NN search with prefiltering, on faiss index with PQ

### ISSUE THREAD

**Issue by tomekcejner-cloudinary** [opensearch-project/k-NN#1057]
**What is the bug?**
I ran into specific use case, the index is created using trained model for PQ encoder, and while regular k-NN search seem to work, when I make use of pre-filtering (introduced in 2.9 release), the results are empty. No error is reported.

Note that problem occurs only on query with pre-filter. When pre-filter clause is removed from query, then we get results in response.

However, the logs say:

```
[2023-08-23T20:38:06,509][ERROR][o.o.k.i.q.KNNWeight      ] [fa696978b96b] Error while getting the doc values to do the k-NN Search for query : 
java.lang.IllegalArgumentException: Unable to find space: null
	at org.opensearch.knn.index.SpaceType.getSpace(SpaceType.java:132) ~[opensearch-knn-2.9.0.0.jar:2.9.0.0]
	at org.opensearch.knn.index.query.KNNWeight.doExactSearch(KNNWeight.java:293) [opensearch-knn-2.9.0.0.jar:2.9.0.0]
	at org.opensearch.knn.index.query.KNNWeight.scorer(KNNWeight.java:117) [opensearch-knn-2.9.0.0.jar:2.9.0.0]
	at org.apache.lucene.search.Weight.bulkScorer(Weight.java:166) [lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.opensearch.search.internal.ContextIndexSearcher$1.bulkScorer(ContextIndexSearcher.java:368) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.search.internal.ContextIndexSearcher.searchLeaf(ContextIndexSearcher.java:313) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:276) [opensearch-2.9.0.jar:2.9.0]
	at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:551) [lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.opensearch.search.query.QueryPhase.searchWithCollector(QueryPhase.java:354) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.search.query.QueryPhase$DefaultQueryPhaseSearcher.searchWithCollector(QueryPhase.java:441) [opensearch-2.9.0.jar:2.9.0]
```

**How can one reproduce the bug?**
Search is performed on 512d vectors, trained via following request:

```json
{
	"training_index": "training-data",
	"training_field": "embedding_vector",
	"dimension": 512,
	"description": "My model",
	"method": {
		"name": "ivf",
		"engine": "faiss",
		"space_type": "l2",
		"parameters": {
			"encoder": {
				"name": "pq",
				"parameters": {
					"code_size": 8,
					"m": 8
				}
			},
			"nlist": 128,
			"nprobes": 128
		}
	}
}
```

And using the index definition:

```json
{
  "settings": {
    "index": {
      "knn": true
    },
    "number_of_shards": 10,
    "number_of_replicas": 1
  },
  "mappings": {
    "properties": {
      "image_id": {
        "type": "keyword"
      },
      "customer_id": {
        "type": "keyword"
      },
      "embedding_vector": {
        "type": "knn_vector",
        "dimension": 512,
         "model_id": "ivf_pq_3"
      }
    }
  }
}
```

And following query causes the error visible only in the logs.

```json
{
  "size": 50,
  "query": {
    "knn": {
      "embedding_vector": {
        "filter": {
          "bool": {
            "must": [
              {
                "term": {
                  "customer_id": 1
                }
              }
            ]
          }
        },
        "vector" : [...],
        "k": 50
      }
    }
  }
}
				
```

**What is the expected behavior?**
Non-empty search results, or error message, if there's actually an error in usage.

**Comment by tomekcejner-cloudinary** [comment]
@naveentatikonda out of curiosity, can you confirm it this is indeed a bug?

**Comment by navneet1v** [comment]
@tomekcejner-cloudinary the filters(Efficient Filters) with IVF PQ is not supported in 2.9. It would be supported with version 2.10 of OpenSearch. I backported the change till 2.9 version to make sure that if 2.9.1 version is released then support for filtering comes with IVF-PQ.

**Comment by navneet1v** [comment]
The changes for this issue is merged in 2.10 branch and will be released with 2.10 version of OpenSearch. Will close the issue once the release is done.

**Comment by navneet1v** [comment]
> @tomekcejner-cloudinary the filters(Efficient Filters) with IVF PQ is not supported in 2.9. It would be supported with version 2.10 of OpenSearch. I backported the change till 2.9 version to make sure that if 2.9.1 version is released then support for filtering comes with IVF-PQ.

The same is mentioned in the documentation too: https://opensearch.org/docs/latest/search-plugins/knn/filter-search-knn/#:~:text=Faiss%20engine%20with%20an%20HNSW%20algorithm%20(k%2DNN%20plugin%20versions%202.9%20or%20later)

**Comment by navneet1v** [comment]
Resolving this issue as the code changes are merged and will be release with 2.10 version of OpenSearch.
